nemo/collections/asr/parts/numba/rnnt_loss/utils/cuda_utils/gpu_rnnt_kernel.py

```diff
@@ -397,3 +397,455 @@ def compute_grad_kernel(
+        )
+        for i in range(num_big_blanks):
+            if T >= big_blank_duration[i] and big_blank_duration[i] == 1:
+                betas[offset + (T - 1) * maxU + U - 1] = rnnt_helper.log_sum_exp(
+                    betas[offset + (T - 1) * maxU + U - 1],
+                    logp(denom, acts, maxT, maxU, alphabet_size, b, T - 1, U - 1, blank_ - 1 - i) - sigma,
+                )
+
+    # sync until all betas are initialized
+    cuda.syncthreads()
+
+    # Ordinary beta calculations, broadcast across B=b and U=u
+    # Look up backward variable calculation from rnnt_numpy.backward_pass()
+    for n in range(T + U - 2, -1, -1):
+        t = n - u
+
+        if u == (U - 1):
+            # for t in reversed(range(T - 1)) step to initialize betas[b, t, U-1]
+            if t >= 0 and t < (T - 1):
+                betas[offset + t * maxU + U - 1] = (
+                    betas[offset + (t + 1) * maxU + U - 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, U - 1, blank_)
+                    - sigma
+                )
+
+                # now add the weights from big blanks
+                for i in range(num_big_blanks):
+                    if t + big_blank_duration[i] < T:
+                        betas[offset + t * maxU + U - 1] = rnnt_helper.log_sum_exp(
+                            betas[offset + t * maxU + U - 1],
+                            betas[offset + (t + big_blank_duration[i]) * maxU + U - 1]
+                            + logp(denom, acts, maxT, maxU, alphabet_size, b, t, U - 1, blank_ - 1 - i)
+                            - sigma,
+                        )
+                    elif t + big_blank_duration[i] == T and big_blank_duration[i] != 1:
+                        betas[offset + t * maxU + U - 1] = rnnt_helper.log_sum_exp(
+                            betas[offset + t * maxU + U - 1],
+                            logp(denom, acts, maxT, maxU, alphabet_size, b, t, U - 1, blank_ - 1 - i) - sigma,
+                        )
+
+        elif u < U:
+            if t == T - 1:
+                # for u in reversed(range(U - 1)) step to initialize betas[b, T-1, u]
+                betas[offset + (T - 1) * maxU + u] = (
+                    betas[offset + (T - 1) * maxU + u + 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, T - 1, u, labels[u])
+                    - sigma
+                )
+            elif (t >= 0) and (t < T - 1):
+                # for t in reversed(range(T - 1)) for u in reversed(range(U - 1)) step to compute betas[b, t, u]
+                no_emit = (
+                    betas[offset + (t + 1) * maxU + u]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, u, blank_)
+                    - sigma
+                )
+                emit = (
+                    betas[offset + t * maxU + u + 1]
+                    + logp(denom, acts, maxT, maxU, alphabet_size, b, t, u, labels[u])
+                    - sigma
```

Comment:
Two able

Reply:
What's two able?

Reply:
Hah i meant refer to two above. Phone is useless for reviews sometimes.